### ISSUE TITLE
"error: attempt to use a deleted function" when compiling minimal example in MacOS Sierra

### ISSUE THREAD

**Issue by LordGolias** [GameFoundry/bsf#167]
Steps to reproduce:

1. install latest cmake with brew
2. Download and untar the binaries for mac to `/Users/u/Downloads/bsf_v1.0.0b_osx`
3. Create new directory and copy `Findbsf.cmake`
4. Create `CMakeLists.txt` with 

```
# Minimum version of CMake as required by bsf
cmake_minimum_required (VERSION 3.6.0)
# Name of your project
project (myProject)
# Add necessary compiler flags
if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR
   CMAKE_CXX_COMPILER_ID MATCHES "AppleClang" OR
   CMAKE_CXX_COMPILER_ID MATCHES "GNU")
    add_compile_options(-fno-rtti)
endif()

# Build an executable from the provided C++ files
add_executable(myApp "graphics.cpp")

# Path to where you have installed bsf
set(bsf_INSTALL_DIR "/Users/u/Downloads/bsf_v1.0.0b_osx")

# Let CMake know where to find the Findbsf.cmake file (at current folder)
set(CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}")

# Find bsf libraries and headers
find_package(bsf REQUIRED)

# Link bsf library with your application
target_link_libraries(myApp PRIVATE bsf)
```

4. Create `graphics.cpp` with the minimal example:

```
#include "BsApplication.h"

int main()
{
    using namespace bs;
    VideoMode videoMode(1280, 720);
    Application::startUp(videoMode, "Example", false);
    Application::instance().runMainLoop();
    Application::shutDown();
    return 0;
}
```

5. compile the project (e.g. CLion)

This causes the following error

```
In file included from /Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfUtility/Prerequisites/BsPrerequisitesUtil.h:262:
/Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfUtility/Allocators/BsMemoryAllocator.h:427:50: error: attempt to use a deleted function
                static constexpr void destroy(pointer p) { p->~T(); }
                                                               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1357:21: note: in instantiation of member function 'bs::StdAlloc<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, void *>, bs::GenAlloc>::destroy' requested here
    -> decltype(__a.destroy(__p), true_type());
                    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1368:22: note: while substituting deduced template arguments into function template '__has_destroy_test' [with _Alloc = bs::StdAlloc<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, void *>, bs::GenAlloc>, _Pointer = std::__1::pair<const std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > > *]
            decltype(__has_destroy_test(declval<_Alloc>(),
                     ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1548:24: note: in instantiation of template class 'std::__1::__has_destroy<bs::StdAlloc<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, void *>, bs::GenAlloc>, std::__1::pair<const std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > > *>' requested here
            {__destroy(__has_destroy<allocator_type, _Tp*>(), __a, __p);}
                       ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:1793:24: note: in instantiation of function template specialization 'std::__1::allocator_traits<bs::StdAlloc<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, void *>, bs::GenAlloc> >::destroy<std::__1::pair<const std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > > >' requested here
        __node_traits::destroy(__na, _NodeTypes::__get_ptr(__nd->__value_));
                       ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:1781:5: note: in instantiation of member function 'std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, true>, bs::StdAlloc<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, bs::GenAlloc> >::destroy' requested here
    destroy(__root());
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/map:807:29: note: in instantiation of member function 'std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, true>, bs::StdAlloc<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, bs::GenAlloc> >::~__tree' requested here
class _LIBCPP_TYPE_VIS_ONLY map
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:757:3: note: '~__tree_node' has been explicitly marked deleted here
  ~__tree_node() _LIBCPP_EQUAL_DELETE;
  ^
In file included from /Users/u/ClionProjects/tactical_ai/examples/3_graphics/graphics.cpp:1:
In file included from /Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfEngine/BsApplication.h:5:
In file included from /Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfEngine/BsPrerequisites.h:5:
In file included from /Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfCore/BsCorePrerequisites.h:5:
In file included from /Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfUtility/Prerequisites/BsPrerequisitesUtil.h:262:
/Users/u/Downloads/bsf_v1.0.0b_osx/include/bsfUtility/Allocators/BsMemoryAllocator.h:427:50: error: '~__tree_node' is a private member of 'std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> >, std::__1::basic_string<char, std::__1::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > >, void *>'
                static constexpr void destroy(pointer p) { p->~T(); }
                                                               ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tree:757:3: note: declared private here
  ~__tree_node() _LIBCPP_EQUAL_DELETE;
  ^
2 errors generated.
make[3]: *** [examples/3_graphics/CMakeFiles/myApp.dir/graphics.cpp.o] Error 1
make[2]: *** [examples/3_graphics/CMakeFiles/myApp.dir/all] Error 2
make[1]: *** [examples/3_graphics/CMakeFiles/myApp.dir/rule] Error 2
make: *** [myApp] Error 2
```

Any ideas of what this might be?

**Comment by BearishSun** [comment]
I'll take a closer look when I get more time, not on mac currently. But for now, what is your XCode and Clang version?

**Comment by LordGolias** [comment]
Xcode: `Version 8.3.3 (8E3004b)` (via About in Xcode)
CLang: `Apple LLVM version 8.1.0 (clang-802.0.42)` (via `clang --version`)

**Comment by lysannschlegel** [comment]
FYI: I had the same issue and fixed it by using Xcode 9.2 instead of Xcode 8.3.3.

**Comment by BearishSun** [comment]
Tested right now on Xcode 11, everything appears to compile. Please upgrade Xcode version and re-open issue if the problem isn't resolved.
